src/client/Client.cc

```diff
@@ -14087,6 +14089,11 @@ int Client::_removexattr(Inode *in, const char *name, const UserPerm& perms)
  
   int res = make_request(req, perms);
 
+  if ((!strcmp(name, ACL_EA_ACCESS) ||
+      !strcmp(name, ACL_EA_DEFAULT)) &&
+      res == -CEPHFS_ENODATA)
+    res = 0;
+
```

Comment:
But here, I am afraid you are changing the logic.

In kclient it will switch to `CEPH_MDS_OP_RMXATTR` only when `value == nullptr` && `flags & CEPH_XATTR_REPLACE`:

```
commit 04303d8ad06985bed4ea4f46018fc8f55a0962a8
Author: Yan, Zheng <zyan@redhat.com>
Date:   Thu Apr 21 12:11:54 2016 +0800

    ceph: use CEPH_MDS_OP_RMXATTR request to remove xattr
    
    Setxattr with NULL value and XATTR_REPLACE flag should be equivalent
    to removexattr. But current MDS does not support deleting vxattrs through
    MDS_OP_SETXATTR request. The workaround is sending MDS_OP_RMXATTR request
    if setxattr actually removs xattr.
    
    Signed-off-by: Yan, Zheng <zyan@redhat.com>
``` 

And only this case will the kclient issues `CEPH_MDS_OP_RMXATTR`. 

So I think your app just triggered `CEPH_MDS_OP_SETXATTR` & `flags & CEPH_XATTR_REMOVE` to removexattr. And in this case the MDS will always return success when the corresponding xattr doesn't exist.

While the `CEPH_MDS_OP_SETXATTR` will always return `-ENODATA`.

The userspace client's logic is different and your app seems triggered `CEPH_MDS_OP_SETXATTR` directly.

Reply:
> But here, I am afraid you are changing the logic.
> 
> In kclient it will switch to `CEPH_MDS_OP_RMXATTR` only when `value == nullptr` && `flags & CEPH_XATTR_REPLACE`:
> 
> And only this case will the kclient issues `CEPH_MDS_OP_RMXATTR`.
> 
> So I think your app just triggered `CEPH_MDS_OP_SETXATTR` & `flags & CEPH_XATTR_REMOVE` to removexattr. And in this case the MDS will always return success when the corresponding xattr doesn't exist.
> 
> While the `CEPH_MDS_OP_SETXATTR` will always return `-ENODATA`.

Thanks for taking a detailed look.

> The userspace client's logic is different and your app seems triggered `CEPH_MDS_OP_SETXATTR` directly.

The custom application used _ceph_fremovexattr()_ to remove non-existent POSIX ACL xattrs. Is it then wrong to expect 0 as return code via libcephfs? Currently it returns -ENODATA whereas kclient returns 0.

Reply:
The `ceph_fremovexattr()` will trigger `CEPH_MDS_OP_RMXATTR` in userspace client and `-ENODATA` is expected.

BTW, do you know how does the app do with the kclient exactly ? I guess it just triggers `setxattr` with NULL `value` && `CEPH_XATTR_REPLACE` being absent.

Reply:
> The `ceph_fremovexattr()` will trigger `CEPH_MDS_OP_RMXATTR` in userspace client and `-ENODATA` is expected.

Understood. But is that correct is the question.

> BTW, do you know how does the app do with the kclient exactly ? I guess it just triggers `setxattr` with NULL `value` && `CEPH_XATTR_REPLACE` being absent.

It just calls _fremovexattr()_ and return is a success and thus the question.

Reply:
> > The `ceph_fremovexattr()` will trigger `CEPH_MDS_OP_RMXATTR` in userspace client and `-ENODATA` is expected.
> 
> Understood. But is that correct is the question.

i.e, aren't we missing special handling for POSIX ACL xattrs where an attempt to remove one among them is supposed to return success instead of `-ENODATA` irrespective of their presence? If not in my opinion that's a mismatch in behaviour.

Reply:
> > The `ceph_fremovexattr()` will trigger `CEPH_MDS_OP_RMXATTR` in userspace client and `-ENODATA` is expected.
> 
> Understood. But is that correct is the question.
> 
> > BTW, do you know how does the app do with the kclient exactly ? I guess it just triggers `setxattr` with NULL `value` && `CEPH_XATTR_REPLACE` being absent.
> 
> It just calls _fremovexattr()_ and return is a success and thus the question.

Yeah, IMO we should fix this in MDS to make them consistent.

Reply:
@vshankar What do you think ?

Reply:
Let me know if the current approach needs to be reworked so that the patch set can be modified to hand over _removexattr()_ requests for just POSIX ACL xattrs to MDS. Further modifications to MDS can be handled in a separate PR if required.

Reply:
> > > > BTW, do you know how does the app do with the kclient exactly ? I guess it just triggers `setxattr` with NULL `value` && `CEPH_XATTR_REPLACE` being absent.
> > > 
> > > 
> > > It just calls _fremovexattr()_ and return is a success and thus the question.
> > 
> > 
> > Yeah, IMO we should fix this in MDS to make them consistent.
> 
> I am afraid it is hitting kernel ceph code at all. It is intercepted using _is_posix_acl_xattr()_ inside _removexattr()_ within [fs/xattr.c](https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/tree/fs/xattr.c) which further returns success after necessary permission and security checks are performed.

Hm..but I'm not sure how the presence of POSIX ACL xattr itself is determined though. May be I got it all wrong?